File: modelalive/scan.py
from __future__ import annotations

import re
from dataclasses import dataclass, field
from pathlib import Path

from modelalive.check import alive

# Common LLM model ID patterns in source code and config
_MODEL_PATTERNS = [
    re.compile(r"""['"](claude-[a-zA-Z0-9._-]{3,128})['"]"""),
    re.compile(r"""['"](gpt-[a-zA-Z0-9._-]{3,128})['"]"""),
    re.compile(r"""['"](gemini-[a-zA-Z0-9._-]{3,128})['"]"""),
    re.compile(r"""['"](o[0-9]-[a-zA-Z0-9._-]{3,128})['"]"""),
    re.compile(r"""['"](text-embedding-[a-zA-Z0-9._-]{3,64})['"]"""),
    re.compile(r"""['"](ft:[^'"]{3,200})['"]"""),
    re.compile(r"""['"](llama-[a-zA-Z0-9._-]{3,128})['"]"""),
    re.compile(r"""['"](grok-[a-zA-Z0-9._-]{3,128})['"]"""),
    re.compile(r"""['"](deepseek-[a-zA-Z0-9._-]{3,128})['"]"""),
    re.compile(r"""['"](mistral-[a-zA-Z0-9._-]{3,128})['"]"""),
    re.compile(r"""['"](qwen[a-zA-Z0-9._-]{0,128})['"]"""),
    re.compile(r"""['"](anthropic\.claude-[a-zA-Z0-9._:-]{3,128})['"]"""),
    re.compile(r"""['"](amazon\.[a-zA-Z0-9._:-]{3,128})['"]"""),
    re.compile(
        r"""['"]((?:anthropic|openai|google|meta-llama|qwen|deepseek|mistral|x-ai)/[a-zA-Z0-9._:-]{3,128})['"]"""
    ),
    re.compile(r"""['"](meta-llama/[a-zA-Z0-9._:-]{3,128})['"]"""),
    re.compile(r"""model\s*[=:]\s*['"]([a-zA-Z0-9][a-zA-Z0-9._:/-]{2,127})['"]"""),
]

_SKIP_DIRS = {
    ".git",
    ".venv",
    "venv",
    "node_modules",
    "__pycache__",
    ".pytest_cache",
    "dist",
    "build",
    ".eggs",
}

_SKIP_FILES = {".min.js", ".map"}


@dataclass
class ScanFinding:
    path: str
    line: int
    model: str
    status: str
    replacement: str | None = None
    alive: bool = True


@dataclass
class ScanReport:
    root: str
    scanned_files: int = 0
    findings: list[ScanFinding] = field(default_factory=list)

    @property
    def retired(self) -> list[ScanFinding]:
        return [f for f in self.findings if f.status == "retired"]

    @property
    def deprecated(self) -> list[ScanFinding]:
        return [f for f in self.findings if f.status == "deprecated"]

    @property
    def unknown(self) -> list[ScanFinding]:
        return [f for f in self.findings if f.status == "unknown"]


def _should_scan(path: Path) -> bool:
    if any(part in _SKIP_DIRS for part in path.parts):
        return False
    if path.suffix in {".png", ".jpg", ".jpeg", ".gif", ".ico", ".woff", ".woff2", ".ttf"}:
        return False
    if any(path.name.endswith(s) for s in _SKIP_FILES):
        return False
    return path.is_file()


def scan_path(
    root: str | Path,
    *,
    extensions: set[str] | None = None,
) -> ScanReport:
    root_path = Path(root).resolve()
    ext = extensions or {
        ".py",
        ".ts",
        ".tsx",
        ".js",
        ".jsx",
        ".json",
        ".yaml",
        ".yml",
        ".toml",
        ".env",
        ".md",
        ".sh",
    }
    report = ScanReport(root=str(root_path))
    seen: set[tuple[str, int, str]] = set()

    for path in root_path.rglob("*"):
        if not _should_scan(path):
            continue
        if path.suffix not in ext and path.name not in {"Dockerfile", ".env"}:
            continue
        try:
            text = path.read_text(encoding="utf-8", errors="ignore")
        except OSError:
            continue
        report.scanned_files += 1
        for line_no, line in enumerate(text.splitlines(), start=1):
            for pattern in _MODEL_PATTERNS:
                for match in pattern.finditer(line):
                    model = match.group(1)
                    key = (str(path.relative_to(root_path)), line_no, model)
                    if key in seen:
                        continue
                    seen.add(key)
                    try:
                        result = alive(model)
                    except ValueError:
                        continue
                    if result.status == "active":
                        continue
                    report.findings.append(
                        ScanFinding(
                            path=key[0],
                            line=line_no,
                            model=model,
                            status=result.status,
                            replacement=result.replacement,
                            alive=result.alive,
                        )
                    )
    return report

File: js/src/scan.ts
import { alive } from "./alive.js";
import type { AliveResult } from "./types.js";
import { readdirSync, readFileSync, statSync } from "node:fs";
import { join } from "node:path";

const SKIP_DIRS = new Set([
  ".git",
  ".venv",
  "venv",
  "node_modules",
  "__pycache__",
  ".pytest_cache",
  "dist",
  "build",
  ".eggs",
]);

const EXTENSIONS = new Set([
  ".py",
  ".ts",
  ".tsx",
  ".js",
  ".jsx",
  ".json",
  ".yaml",
  ".yml",
  ".toml",
  ".env",
  ".md",
  ".sh",
]);

const PATTERNS: RegExp[] = [
  /['"](claude-[a-zA-Z0-9._-]{3,128})['"]/g,
  /['"](gpt-[a-zA-Z0-9._-]{3,128})['"]/g,
  /['"](gemini-[a-zA-Z0-9._-]{3,128})['"]/g,
  /['"](o[0-9]-[a-zA-Z0-9._-]{3,128})['"]/g,
  /['"](llama-[a-zA-Z0-9._-]{3,128})['"]/g,
  /['"](grok-[a-zA-Z0-9._-]{3,128})['"]/g,
  /['"](deepseek-[a-zA-Z0-9._-]{3,128})['"]/g,
  /['"](mistral-[a-zA-Z0-9._-]{3,128})['"]/g,
  /['"](qwen[a-zA-Z0-9._-]{0,128})['"]/g,
  /['"](anthropic\.claude-[a-zA-Z0-9._:-]{3,128})['"]/g,
  /['"]((?:anthropic|openai|google|meta-llama|qwen|deepseek|mistral|x-ai)\/[a-zA-Z0-9._:-]{3,128})['"]/g,
  /model\s*[=:]\s*['"]([a-zA-Z0-9][a-zA-Z0-9._:/-]{2,127})['"]/g,
];

export interface ScanFinding {
  path: string;
  line: number;
  model: string;
  status: string;
  replacement?: string | null;
  alive: boolean;
}

export interface ScanReport {
  root: string;
  scannedFiles: number;
  findings: ScanFinding[];
}

function walk(dir: string, files: string[] = []): string[] {
  for (const name of readdirSync(dir)) {
    const path = join(dir, name);
    const parts = path.split(/[/\\]/);
    if (parts.some((p) => SKIP_DIRS.has(p))) continue;
    const st = statSync(path);
    if (st.isDirectory()) walk(path, files);
    else files.push(path);
  }
  return files;
}

function shouldScan(path: string): boolean {
  if (path.endsWith(".min.js") || path.endsWith(".map")) return false;
  const ext = path.slice(path.lastIndexOf("."));
  return EXTENSIONS.has(ext) || path.endsWith("Dockerfile") || path.endsWith(".env");
}

export function scanPath(root: string, today = new Date()): ScanReport {
  const report: ScanReport = { root, scannedFiles: 0, findings: [] };
  const seen = new Set<string>();

  for (const path of walk(root)) {
    if (!shouldScan(path)) continue;
    let text: string;
    try {
      text = readFileSync(path, "utf-8");
    } catch {
      continue;
    }
    report.scannedFiles += 1;
    const lines = text.split(/\r?\n/);
    for (let i = 0; i < lines.length; i++) {
      const line = lines[i]!;
      for (const pattern of PATTERNS) {
        pattern.lastIndex = 0;
        let match: RegExpExecArray | null;
        while ((match = pattern.exec(line)) !== null) {
          const model = match[1]!;
          const key = `${path}:${i + 1}:${model}`;
          if (seen.has(key)) continue;
          seen.add(key);
          let result: AliveResult;
          try {
            result = alive(model, today);
          } catch {
            continue;
          }
          if (result.status === "active") continue;
          report.findings.push({
            path: path.startsWith(root) ? path.slice(root.length + 1) : path,
            line: i + 1,
            model,
            status: result.status,
            replacement: result.replacement,
            alive: result.alive,
          });
        }
      }
    }
  }
  return report;
}
